### ISSUE TITLE
Using the new native M1 support

### ISSUE THREAD

**Issue by purcell** [LnL7/nix-darwin#334]
Apple M1 (`aarch64-darwin`) support just landed in Nixpkgs `master`, so I thought I'd open an issue here to describe/discuss how best to migrate to a `nix-darwin` setup which uses that system type by default.

I've had some luck with this, so I'll recount what I've done. Starting with a regular `x86_64-darwin` installation, on the `nixpkgs-unstable` channel, I added
```
      extra-platforms = x86_64-darwin aarch64-darwin
```
to `nix.extraOptions` and defined a package set like
```
  pkgsM1 = import <nixpkgs> { localSystem = "aarch64-darwin"; };
```

At this point, it's possible to set `nix.pkg = pkgsM1.nix`, and then after `darwin-rebuild switch`, the nix tools and overall system are switched to aarch64, and `pkgs` will yield `aarch64` packages.

However, of course many packages in a user's configuration won't build on `aarch64`, so to get to this stage, you have to comment out those packages. After switching the nix installation to `aarch64`, the packages can be added back in, with a similar `pkgs_x86`.

Notable packages (for me) which don't yet build on `aarch64` are `nix-direnv` and anything written in Haskell, like `shellcheck`, `niv`.

It's also possible to *skip* switching `nix.pkg`, and keep the overall system as `x86_64` for now: then, packages can be included from `pkgsM1` for installation as desired. This might be the best choice for most users until some of the rough edges are smoothed off.

Hope this helps some people. I'd be keen to hear tips and tricks from others.

**Comment by hlolli** [comment]
Thanks a lot for this tip. I've finally got the wheels running after mostly commenting out package overrides based on nixpkgs PR's. I went the other way around it. Instead of selecting the aarch64 packages, I overlay the x86_64's.

`~/.nixpkgs/darwin-configuration.nix`

I use with to thoroughly set the stdenv.
```nix
{ ... }:
# just a tarball of the latest master
with import (fetchTarball /. + "file:///Users/hlolli/nixpkgs-fork.tar.gz") { localSystem = "aarch64-darwin"; };

 {
  imports = [
    /Users/hlolli/nixos-configuration
  ];
}
```

`/Users/hlolli/nixos-configuration/default.nix`

I have the two versions
```nix
{ stdenv, config, pkgs, ... }:

let pkgsX86 = import <nixpkgs> { localSystem = "x86_64-darwin"; };
...
```

and overlay the unsupported packages

```nix
...
  nixpkgs.overlays = [
    (
    self: super: {
      inherit (pkgsX86) haskell haskellPackages openssl openssl_1_1;
    }
  )];
  ...
  ```
  
  long rebuild ahead rn, praying for luck 🙏

**Comment by LnL7** [comment]
I don't think that overlay will do what you want as it mixes x86 and arm libraries. To build a mixed system you most likely want to use the same overlays for both package sets and between the architecture at binary level.

```nix
{ config, lib, pkgs, ... }:
let pkgsX86 = import <nixpkgs> { localSystem = "x86_64-darwin"; overlays = config.nixpkgs.overlays; };
{
  environment.systemPackages =
    [ pkgs.curl
      pkgsX86.haskell
    ];
}
```

**Comment by hlolli** [comment]
Yes indeed, I should have known, but the rosetta2 magic has spelled my senses here.
Bad idea I was suggesting there above.

```
ld: warning: ignoring file /nix/store/96j2ai9h4f5kaqcnp9kvgbp2srr79xhx-openssl-1.1.1k/lib/libcrypto.dylib, building for macOS-arm64 but attempting to link with file built for macOS-x86_64
Undefined symbols for architecture arm64:
```

Besides the fact that openssl compiles fine on aarch (so far so good after I removed it from the overlay), then I hope this adverts pitfalls for others.

**Comment by hlolli** [comment]
Maybe not the right place for it, but I have a PR for emacs fix for aarch64, if someone wants to review and test the fix https://github.com/NixOS/nixpkgs/pull/124888

**Comment by antifuchs** [comment]
I've gotten a flake.nix-based configuration working working, where I install the rosetta x86-64 packages by default and opt into amd64 packages on a case-by-case basis.

Here's the relevant flake.nix bits:

``` nix
outputs {self, nixpkgs, ...}: {
  darwinConfigs."my-machine" = darwin.lib.darwinSystem {
    modules = [
      { nix.extraOptions = ''extra-platforms = x86_64-darwin aarch64-darwin ''; }
      ./machine/my-machine.nix
    ];
    specialArgs = { 
      m1Ize = (pkgs: config: import nixpkgs {
        system = "aarch64-darwin";
        overlays = config.nixpkgs.overlays;
      });
    };
  };
}
```

note the `specialArgs.m1Ize` function; that will be passed down to the modules, e.g. to my development system config module:

``` nix
{ pkgs, config, lib, m1Ize, ... }:
let
  m1Pkgs = m1Ize pkgs config;
in
{
  environment.systemPackages = with m1Pkgs; [
    cmake
    llvm
    openssl
    gnumake
    postgresql_13
  ];
}
```

You have to use that `let` binding in all the modules where you intend to use it, but it's about as short as it gets. Hope this helps fellow flake-ers (:

**Comment by lloeki** [comment]
For those like me that use a pure nixpkgs install (i.e don't use nix-darwin, flakes, or whatever), I'm popping up a `nix-shell` with arm64 derivations quite easily:

```
$ sh <(curl -L https://nixos.org/nix/install) --daemon  # install as usual
$ cat /etc/nix/nix.conf                                 # add extra-platforms

build-users-group = nixbld
extra-platforms = x86_64-darwin aarch64-darwin
$ cat shell.nix
{
  pkgs ? import <nixpkgs> { localSystem = "aarch64-darwin"; },
}:
let
  ruby = pkgs.ruby_3_0;
in pkgs.mkShell {
  buildInputs = [
    pkgs.openssl
    ruby
  ];
}
$ nix-shell
nix$ which ruby
/nix/store/sc54smx4gqa80nxwpfvpxfw8slh7fgyd-ruby-3.0.1/bin/ruby
nix$ file $(which ruby)
/nix/store/sc54smx4gqa80nxwpfvpxfw8slh7fgyd-ruby-3.0.1/bin/ruby: Mach-O 64-bit executable arm64
```

Now I'm a complete beginner, so if only I knew how to pivot the nixpkgs install into arm64 😅

**Comment by LnL7** [comment]
@lloeki By default the nix client evaluates using the system it was built with. So upgrading you nix installation to the arm build will also change the default.

**Comment by lloeki** [comment]
Thanks @LnL7, I think I managed to do it:

```
$ sudo su -l
# cat /etc/nix/nix.conf 
build-users-group = nixbld
system = aarch64-darwin
extra-platforms = x86_64-darwin aarch64-darwin
#  launchctl unload /Library/LaunchDaemons/org.nixos.nix-daemon.plist
#  launchctl load /Library/LaunchDaemons/org.nixos.nix-daemon.plist
# nix-env -iA nixpkgs.nix
replacing old 'nix-2.3.12'
installing 'nix-2.3.12'
-----8<------
# file $(which nix-shell)
/var/root/.nix-profile/bin/nix-shell: Mach-O 64-bit executable arm64
# file /nix/var/nix/profiles/default/bin/nix-daemon
/nix/var/nix/profiles/default/bin/nix-daemon: Mach-O 64-bit executable arm64
#  launchctl unload /Library/LaunchDaemons/org.nixos.nix-daemon.plist
#  launchctl load /Library/LaunchDaemons/org.nixos.nix-daemon.plist
```

From there I think I can remove `system = aarch64-darwin` from `/etc/nix/nix.conf`, but everything seems alright nonetheless.

**Comment by shaunsingh** [comment]
Hi, I'm having an error with installing `nix-darwin` using my configuration.nix. My configuration.nix can be found here: https://github.com/shaunsingh/vimrc-dotfiles/blob/main/.nixpkgs/darwin-configuration.nix

Please let me know if you need any other information that may help with solving the issue. Thank you for the amazing project and your time!

This is the error produced: 
```zsh
~
❯ nix-build https://github.com/LnL7/nix-darwin/archive/master.tar.gz -A installer
  ./result/bin/darwin-installer

/nix/store/3dkzr2vp4yfpizps11vmjxzm2x8j8pnv-darwin-installer

Installing nix-darwin...

Would you like to edit the default configuration.nix before starting? [y/n] y
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels' does not exist, ignoring
error: while evaluating the attribute 'activationScript' of the derivation 'darwin-system-21.11pre295944.0747387223e+darwin4.0000000' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/pkgs/stdenv/generic/make-derivation.nix:201:11:
while evaluating the attribute 'activationScripts.script.text' at undefined position:
while evaluating 'g' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/attrsets.nix:298:19, called from undefined position:
while evaluating anonymous function at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:140:72, called from /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/attrsets.nix:301:20:
while evaluating the attribute 'value' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:525:9:
while evaluating the option `system.activationScripts.script.text':
while evaluating the attribute 'mergedValue' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:557:5:
while evaluating the attribute 'values' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:551:9:
while evaluating the attribute 'values' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:650:7:
while evaluating anonymous function at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:537:28, called from /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:537:17:
while evaluating definitions from `/nix/store/pafi8wjxc8f9sslyb578cwddsrpki1nv-nix-darwin/modules/system/activation-scripts.nix':
while evaluating 'dischargeProperties' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:609:25, called from /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:538:137:
while evaluating the attribute 'value' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:392:44:
while evaluating the attribute 'activationScripts.applications.text' at undefined position:
while evaluating 'g' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/attrsets.nix:298:19, called from undefined position:
while evaluating anonymous function at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:140:72, called from /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/attrsets.nix:301:20:
while evaluating the attribute 'value' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:525:9:
while evaluating the option `system.activationScripts.applications.text':
while evaluating the attribute 'mergedValue' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:557:5:
while evaluating the attribute 'values' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:551:9:
while evaluating the attribute 'values' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:650:7:
while evaluating anonymous function at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:537:28, called from /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:537:17:
while evaluating definitions from `/nix/store/pafi8wjxc8f9sslyb578cwddsrpki1nv-nix-darwin/modules/system/applications.nix':
while evaluating 'dischargeProperties' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:609:25, called from /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:538:137:
while evaluating the attribute 'value' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:392:44:
while evaluating the attribute 'passAsFile' of the derivation 'system-applications' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/pkgs/stdenv/generic/make-derivation.nix:201:11:
while evaluating the attribute 'passAsFile' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/pkgs/build-support/buildenv/default.nix:77:5:
while evaluating the attribute 'environment.systemPackages' at undefined position:
while evaluating 'g' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/attrsets.nix:298:19, called from undefined position:
while evaluating anonymous function at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:140:72, called from /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/attrsets.nix:301:20:
while evaluating the attribute 'value' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:525:9:
while evaluating the option `environment.systemPackages':
while evaluating the attribute 'mergedValue' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:557:5:
while evaluating 'merge' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/types.nix:357:20, called from /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:559:59:
while evaluating anonymous function at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/types.nix:358:35, called from undefined position:
while evaluating anonymous function at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/lists.nix:116:29, called from undefined position:
while evaluating anonymous function at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/types.nix:359:21, called from /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/lists.nix:116:32:
while evaluating the attribute 'optionalValue' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:569:5:
while evaluating the attribute 'values' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:551:9:
while evaluating the attribute 'values' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:650:7:
while evaluating anonymous function at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:537:28, called from /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:537:17:
while evaluating definitions from `/Users/shauryasingh/.nixpkgs/darwin-configuration.nix':
while evaluating 'dischargeProperties' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:609:25, called from /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/modules.nix:538:137:
while evaluating the attribute 'value' at /nix/store/sqb7kisc8x4q2byphha06vazzw6s3g8s-nixpkgs/lib/types.nix:363:38:
while evaluating the attribute 'packages.aarch64-darwin.neovim' at /nix/store/m3qzm0c5r708l5h8gn7dwrlciib37x0d-source/default.nix:73:15:
attribute 'aarch64-darwin' missing, at /nix/store/r76px2qwq1i8zgfim4angpjkwdd6vh6a-source/flake.nix:16:26
```

**Comment by BaoPham92** [comment]
> Thanks @LnL7, I think I managed to do it:
> 
> ```
> $ sudo su -l
> # cat /etc/nix/nix.conf 
> build-users-group = nixbld
> system = aarch64-darwin
> extra-platforms = x86_64-darwin aarch64-darwin
> #  launchctl unload /Library/LaunchDaemons/org.nixos.nix-daemon.plist
> #  launchctl load /Library/LaunchDaemons/org.nixos.nix-daemon.plist
> # nix-env -iA nixpkgs.nix
> replacing old 'nix-2.3.12'
> installing 'nix-2.3.12'
> -----8<------
> # file $(which nix-shell)
> /var/root/.nix-profile/bin/nix-shell: Mach-O 64-bit executable arm64
> # file /nix/var/nix/profiles/default/bin/nix-daemon
> /nix/var/nix/profiles/default/bin/nix-daemon: Mach-O 64-bit executable arm64
> # launchctl unload /Library/LaunchDaemons/org.nixos.nix-daemon.plist
> # launchctl load /Library/LaunchDaemons/org.nixos.nix-daemon.plist
> ```
> 
> From there I think I can remove `system = aarch64-darwin` from `/etc/nix/nix.conf`, but everything seems alright nonetheless.

I followed both of your above approach while being on macOS Big Sur.

So far I was able to follow the steps, but activating "nix-shell" within the plutus project would give me the following: 

```
error: cannot coerce null to a string, at /nix/store/sdjxjsd5phr225rs2qzklj2xci0c9gr0-source/pkgs/stdenv/generic/make-derivation.nix:192:19
```

I tried: 

```
nix-shell -I nixpkgs=channel:nixpkgs-unstable -p agda
```

From a suggested link here with a related issue for the current [error here](https://github.com/NixOS/nixpkgs/issues/55964#issuecomment-629311290).

Which gave me the following output:

```
error: Package ‘ghc-8.10.4’ in /nix/store/9jyvfa211fxvwyd79vz5ydak07q52y5w-nixexprs.tar.xz/pkgs/development/compilers/ghc/8.10.4.nix:267 is not supported on ‘aarch64-darwin’, refusing to evaluate.

a) To temporarily allow packages that are unsupported for this system, you can use an environment variable
   for a single invocation of the nix tools.

     $ export NIXPKGS_ALLOW_UNSUPPORTED_SYSTEM=1

b) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.
```

I have tried each options of a - c, so far the same error reappears.

I've also attempted your approach as clean installs.

**Comment by yisraeldov** [comment]
@purcell , Steve what is your current setup for getting things working on the M1 ?

**Comment by domenkozar** [comment]
Note that ghc builds as of today on master. More package fixes are pending in haskell-updates (including cachix).

**Comment by domenkozar** [comment]
For flake.nix it needs to pass the correct system to https://github.com/LnL7/nix-darwin/blob/a7492a8c76dcc702d0a65cd820a5f9baa8702684/eval-config.nix#L1

**Comment by dhess** [comment]
https://github.com/LnL7/nix-darwin/pull/317

**Comment by shaunsingh** [comment]
Anyone have a sample `darwin-configuration.nix` setup for m1 that they would like to share? Its hard to follow along with the thread, and I'm quite new to nix-darwin

**Comment by lilyball** [comment]
In order to get this working I had to `darwin-rebuild switch` with `extra-platforms = x86_64-darwin aarch64-darwin` set first, before I could then build any x86_64 packages (`pkgs` is aarch64-darwin). Is there any way to avoid this two-phase configuration? I've got this machine setup now, but I don't want to run into this again the next time I set up a new machine.

**Comment by srid** [comment]
> Anyone have a sample `darwin-configuration.nix` setup for m1 that they would like to share? Its hard to follow along with the thread, and I'm quite new to nix-darwin

@shaunsingh Add this to your `darwin-configuration.nix` during install:

```nix
  nix.extraOptions = ''
    extra-platforms = aarch64-linux
    experimental-features = nix-command flakes
  '';
```

**Comment by masaeedu** [comment]
@domenkozar in your comment [here](https://github.com/NixOS/nixpkgs/issues/95903#issuecomment-914061347) you mention that:

> GHC has been fixed just last week, with now over 25000 packages building on aarch64-darwin.

Presumably this fix is for the "cannot coerce null to a string" bug. Would you happen to recall what PR/commit introduced this fix and what the problem was (I'm wondering if I can backport the fix to an older nixpkgs). Thanks in advance.

**Comment by domenkozar** [comment]
This was related to signing and process getting killed, so your error message is something different/new.

**Comment by yacinehmito** [comment]
Is there still an issue with native M1 support? I installed nix-darwin recently on an M1 and I seem to get everything on aarch64-darwin without having done anything special.

**Comment by purcell** [comment]
Agree, I'll go ahead and close this: I opened it mainly to share info, and there's indeed no current issue on these machines.
